pkg/endpoint/policy.go

```diff
@@ -353,15 +363,71 @@ func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *LabelsMap,
 				e.cleanUnusedRedirects(owner, e.L4Policy.Ingress, c.L4Policy.Ingress)
 				e.cleanUnusedRedirects(owner, e.L4Policy.Egress, c.L4Policy.Egress)
 			}
-			l4Rm, l4Add, err = e.applyL4PolicyLocked(owner, labelsMap, e.L4Policy, c.L4Policy)
+			rulesAdd, l4Rm, err = e.applyL4PolicyLocked(owner, labelsMap, e.L4Policy, c.L4Policy)
 			if err != nil {
 				// This should not happen, and we can't fail at this stage anyway.
-				e.getLogger().Fatal("L4 Policy application failed")
+				e.getLogger().WithError(err).Fatal("L4 Policy application failed")
 			}
 		}
-		e.L4Policy = c.L4Policy // Reuse the common policy
+		// Reuse the common policy, will be used in lxc_config.h (CFG_L4_INGRESS and CFG_L4_EGRESS)
+		e.L4Policy = c.L4Policy
 		e.LabelsMap = labelsMap // Remember the set of labels used
+
+		// We need to know which ones are L4-only by checking
+		// if there's any L4-only rules that doesn't match a L3-L4 rule.
+		if c != nil && c.L4Policy != nil && c.L4Policy.Ingress != nil {
+			for _, l4Filter := range c.L4Policy.Ingress {
+				found := false
+				l4RuleCtx, l7RuleCtx := policy.ParseL4Filter(l4Filter)
+				for _, l4RuleContexts := range rulesAdd {
+					if _, found = l4RuleContexts[l4RuleCtx]; found {
+						break
+					}
+				}
+				if !found {
+					if _, ok := rulesAdd[policy.InvalidIdentity]; !ok {
+						rulesAdd[policy.InvalidIdentity] = policy.NewL4RuleContexts()
+					}
+					l7RuleCtx.L4Installed = true
+					rulesAdd[policy.InvalidIdentity][l4RuleCtx] = l7RuleCtx
+				}
+			}
+		}
+
 		changed = true
```

Comment:
Could this move above the loop above? It's a little out of place since it doesn't seem to impact the loops directly

Reply:
Moved